File: fsharp/f0/f0/Sinks.fs
namespace f0

[<AbstractClass>]
type Sink() =
  abstract member WriteBit: bool -> unit
  abstract member WriteByte:    byte -> unit
  abstract member Finished: unit -> unit
  abstract member State: Reg
  member this.Using (f: Sink -> 'a): 'a =
    let a = f(this)
    this.Finished()
    a

type OutputStreamSink(os: System.IO.Stream) as self = 
    inherit Sink()
    let mutable i = 0
    let mutable buf = Array.create 256 0uy
    let reg = Reg Map.empty
    let mutable bitOffset = 1
    let mutable byteBuf = 0
    override o.WriteByte(b:byte) =
        if bitOffset <> 1
        then
            bitOffset <- 1
            o.WriteByte(byte byteBuf)
            byteBuf <- 0
        if(i=256) then
            os.Write(buf, 0, buf.Length)
            i <- 0
        buf.[i] <- b
        i <- i + 1
    override s.WriteBit (b:bool): unit =
        if(bitOffset = 256)
        then
            bitOffset <- 1
            self.WriteByte(byte byteBuf)
            bitOffset <- 2
            byteBuf <- if b then 1 else 0
        else
            if(b) then byteBuf <- byteBuf ||| bitOffset
            bitOffset <- bitOffset * 2
        ()
    override s.Finished(): unit =
        if bitOffset <> 1 then
            bitOffset <- 1
            self.WriteByte(byte byteBuf)
            byteBuf <- 0
        if i <> 0 then os.Write(buf, 0, i)
        os.Close()
    override s.State: Reg = reg

module Sinks =

    let ToOutputStream(os: System.IO.Stream): Sink = new OutputStreamSink(os) :> Sink
    let ToFile(filename:string): Sink = ToOutputStream(System.IO.File.OpenWrite(filename))
    let ToStdOut() = ToOutputStream(System.Console.OpenStandardOutput())
    let ToStdErr() = ToOutputStream(System.Console.OpenStandardError())

File: scala/src/main/scala/f0/Sinks.scala
package f0

import java.io.FileOutputStream

trait Sink {
  def writeBit(b: Boolean): Unit
  def apply(b: Byte): Unit
  def flush: Unit
  def close: Unit
  def state: Reg
  def using[A](handler: Sink => A): A = try handler(this) finally this.close
  def finishWith[A,F](t: (Writer[A,F], A)) = this.using(s => t._1.bind(s)(t._2))
}

object Sinks {

  def using[A](s:Sink)(handler: Sink => A) = s.using(handler)

  def toFile(filename:String): Sink = toOutputStream(new FileOutputStream(filename))

  def toProcess(p:Process, bufSize: Int = 4096): Sink =
    toOutputStream(new java.io.BufferedOutputStream(p.getOutputStream, bufSize))

  def toOutputStream(os: java.io.OutputStream): Sink = new Sink {
    var i = 0
    var buf = new Array[Byte](256)
    val reg = Reg(Map())
    var bitOffset = 1
    var byteBuf = 0
    def writeBit(b: Boolean): Unit = {
      //println("writing " + b + ", offset: " + bitOffset)
      if (bitOffset == 256) {
        //println("flushing byte buf: " + byteBuf)
        bitOffset = 1
        apply(byteBuf.asInstanceOf[Byte])
        bitOffset = 2
        byteBuf = if (b) 1 else 0
      }
      else {
        if (b) byteBuf |= bitOffset
        bitOffset *= 2
        //println ("buf: " + byteBuf)
      }
    }
    def apply(b: Byte): Unit = {
      if (bitOffset != 1) {
        bitOffset = 1
        apply(byteBuf.asInstanceOf[Byte])
        byteBuf = 0
      }
      if (i == 256) { os.write(buf); i = 0 }
      buf(i) = b; i += 1
    }
    
    def flush: Unit = {
      if (bitOffset != 1) {
        bitOffset = 1
        //println("flushing byte buf on done: " + byteBuf)
        apply(byteBuf.asInstanceOf[Byte])
        byteBuf = 0
      }
      if (i != 0) {
        os.write(buf, 0, i)
        i = 0
      }
    }
    
    def close: Unit = {
      flush
      os.close()
    }
    def state: Reg = reg
  }
}
